build is failing
```
======================================================================
FAIL: * Register house (case type = house, basic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_case_list_form.py", line 161, in test_case_list_form_parent_child_basic
    self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-basic'), factory.app.create_suite())
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 32, in assertXmlEqual
    _check_shared(expected, actual, LXMLOutputChecker(), "xml")
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 126, in _check_shared
    raise AssertionError(message)
AssertionError: XML mismatch
--- want.xml
+++ got.xml
@@ -183,7 +183,7 @@
     <stack>
       <create if="count(instance('commcaresession')/session/data/return_to) = 1 and instance('commcaresession')/session/data/return_to = 'm2' and count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_new_person_0]) &gt; 0">
         <command value="'m2'"/>
-        <datum id="parent_id" value="instance('commcaresession')/session/data/case_id_new_person_0"/>
+        <datum id="parent_id" value="instance('commcaresession')/session/data/case_id"/>
         <datum id="case_id" value="instance('commcaresession')/session/data/case_id_new_person_0"/>
       </create>
       <create if="count(instance('commcaresession')/session/data/return_to) = 1 and instance('commcaresession')/session/data/return_to = 'm2' and count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_new_person_0]) = 0">
----------------------------------------------------------------------
```
```
======================================================================
FAIL: test_link_to_child_module (corehq.apps.app_manager.tests.test_form_workflow.TestFormWorkflow)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_form_workflow.py", line 111, in test_link_to_child_module
    self.assertXmlPartialEqual(self.get_xml('form_link_tdh'), factory.app.create_suite(), "./entry")
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 26, in assertXmlPartialEqual
    self.assertXmlEqual(expected, actual, normalize=False)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 32, in assertXmlEqual
    _check_shared(expected, actual, LXMLOutputChecker(), "xml")
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 126, in _check_shared
    raise AssertionError(message)
AssertionError: XML mismatch
--- want.xml
+++ got.xml
@@ -36,7 +36,7 @@
         <datum id="case_id" value="instance('commcaresession')/session/data/case_id"/>
         <datum id="case_id_new_visit_0" value="uuid()"/>
         <command value="'m2'"/>
-        <datum id="case_id_load_visit_0" value="uuid()"/>
+        <datum id="case_id_load_visit_0" value="instance('commcaresession')/session/data/case_id_new_visit_0"/>
         <command value="'m2-f0'"/>
       </create>
     </stack>
----------------------------------------------------------------------
```

--- a/corehq/apps/app_manager/suite_xml/post_process/workflow.py
+++ b/corehq/apps/app_manager/suite_xml/post_process/workflow.py
@@ -232,7 +232,7 @@ class EndOfFormNavigationWorkflow(object):
                         if child.id not in parent_ids
                     ]
 
-                stack_frames.append(StackFrameMeta(link.xpath, frame_children))
+                stack_frames.append(StackFrameMeta(link.xpath, frame_children, current_session=source_form_datums))
 
         return stack_frames
 
@@ -430,7 +430,18 @@ class StackFrameMeta(object):
     Class used in computing the form workflow.
     """
 
-    def __init__(self, if_clause, children=None, allow_empty_frame=False):
+    def __init__(self, if_clause, children=None, allow_empty_frame=False, current_session=None):
+        """
+        :param if_clause: XPath expression to use in if clause for CreateFrame
+        :type if_clause: string
+        :param children: Child elements for the frame
+        :type children: list
+        :param allow_empty_frame: True if the frame can be empty
+        :type allow_empty_frame: bool
+        :param current_session: List of current session datums
+        :type current_session: list of WorkflowDatumMeta
+        """
+        self.current_session = current_session
         self.if_clause = unescape(if_clause) if if_clause else None
         self.children = []
         self.allow_empty_frame = allow_empty_frame
@@ -448,7 +459,7 @@ class StackFrameMeta(object):
         if not self.children and not self.allow_empty_frame:
             return
 
-        children = _replace_session_references_in_stack(self.children)
+        children = _replace_session_references_in_stack(self.children, current_session=self.current_session)
 
         frame = CreateFrame(if_clause=self.if_clause)
 
@@ -537,7 +548,7 @@ class WorkflowDatumMeta(object):
 session_var_regex = re.compile(r"instance\('commcaresession'\)/session/data/(\w+)")
 
 
-def _replace_session_references_in_stack(stack_children):
+def _replace_session_references_in_stack(stack_children, current_session=None):
     """Given a list of stack children (commands and datums)
     replace any references in the datum to session variables that
     have already been added to the session.
@@ -551,6 +562,7 @@ def _replace_session_references_in_stack(stack_children):
     We have to do this because stack create blocks do not update the session after each datum
     is added so items put into the session in one step aren't available to later steps.
     """
+    current_session_vars = [datum.id for datum in current_session] if current_session else []
     clean_children = []
     child_map = {child.id: child.value for child in stack_children if isinstance(child, StackDatum)}
     for child in stack_children:
@@ -560,7 +572,7 @@ def _replace_session_references_in_stack(stack_children):
         session_vars = session_var_regex.findall(child.value)
         new_value = child.value
         for var in session_vars:
-            if var in child_map:
+            if var in child_map and var not in current_session_vars:
                 new_value = new_value.replace(session_var(var), child_map[var])
 
         child_map[child.id] = new_value

--- a/corehq/apps/app_manager/tests/test_case_list_form.py
+++ b/corehq/apps/app_manager/tests/test_case_list_form.py
@@ -104,14 +104,12 @@ class CaseListFormSuiteTests(SimpleTestCase, TestXmlMixin):
         self.assertXmlEqual(self.get_xml('case-list-form-advanced-autoload'), factory.app.create_suite())
 
     def test_case_list_form_parent_child_advanced(self):
-        """
-        * Register house (case type = house, basic)
-          * Register house form
-        * Register person (case type = person, parent select = 'Register house', advanced)
-          * Register person form
-        * Manager person (case type = person, case list form = 'Register person form', basic)
-          * Manage person form
-        """
+        # * Register house (case type = house, basic)
+        #   * Register house form
+        # * Register person (case type = person, parent select = 'Register house', advanced)
+        #   * Register person form
+        # * Manager person (case type = person, case list form = 'Register person form', basic)
+        #   * Manage person form
         factory = AppFactory(build_version='2.9')
         register_house_module, register_house_form = factory.new_basic_module('register_house', 'house')
         factory.form_opens_case(register_house_form)
@@ -133,14 +131,12 @@ class CaseListFormSuiteTests(SimpleTestCase, TestXmlMixin):
         self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-advanced'), factory.app.create_suite())
 
     def test_case_list_form_parent_child_basic(self):
-        """
-        * Register house (case type = house, basic)
-          * Register house form
-        * Register person (case type = person, parent select = 'Register house', basic)
-          * Register person form
-        * Manager person (case type = person, case list form = 'Register person form', basic)
-          * Manage person form
-        """
+        # * Register house (case type = house, basic)
+        #   * Register house form
+        # * Register person (case type = person, parent select = 'Register house', basic)
+        #   * Register person form
+        # * Manager person (case type = person, case list form = 'Register person form', basic)
+        #   * Manage person form
         factory = AppFactory(build_version='2.9')
         register_house_module, register_house_form = factory.new_basic_module('register_house', 'house')
 
@@ -161,16 +157,14 @@ class CaseListFormSuiteTests(SimpleTestCase, TestXmlMixin):
         self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-basic'), factory.app.create_suite())
 
     def test_case_list_form_parent_child_submodule_basic(self):
-        """
-        * Register house (case type = house, basic)
-          * Register house form
-        * Register person (case type = person, parent select = 'Register house', basic)
-          * Register person form
-        * Update house (case type = house, case list form = 'Register house')
-          * Update house form
-          * Update person (case type = person, case list form = 'Register person form', basic, parent module = 'Update house')
-              * Update person form
-        """
+        # * Register house (case type = house, basic)
+        #   * Register house form
+        # * Register person (case type = person, parent select = 'Register house', basic)
+        #   * Register person form
+        # * Update house (case type = house, case list form = 'Register house')
+        #   * Update house form
+        #   * Update person (case type = person, case list form = 'Register person form', basic, parent module = 'Update house')
+        #       * Update person form
         factory = AppFactory(build_version='2.9')
         register_house_module, register_house_form = factory.new_basic_module('register_house', 'house')
         factory.form_opens_case(register_house_form)
@@ -198,16 +192,14 @@ class CaseListFormSuiteTests(SimpleTestCase, TestXmlMixin):
         self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-submodule-basic'), factory.app.create_suite())
 
     def test_case_list_form_parent_child_submodule_advanced(self):
-        """
-        * Register house (case type = house, basic)
-          * Register house form
-        * Register person (case type = person, parent select = 'Register house', advanced)
-          * Register person form
-        * Update house (case type = house, case list form = 'Register house')
-          * Update house form
-          * Update person (case type = person, case list form = 'Register person form', advanced, parent module = 'Update house')
-              * Update person form
-        """
+        # * Register house (case type = house, basic)
+        #   * Register house form
+        # * Register person (case type = person, parent select = 'Register house', advanced)
+        #   * Register person form
+        # * Update house (case type = house, case list form = 'Register house')
+        #   * Update house form
+        #   * Update person (case type = person, case list form = 'Register person form', advanced, parent module = 'Update house')
+        #       * Update person form
         factory = AppFactory(build_version='2.9')
         register_house_module, register_house_form = factory.new_basic_module('register_house', 'house')
         factory.form_opens_case(register_house_form)
@@ -237,11 +229,10 @@ class CaseListFormSuiteTests(SimpleTestCase, TestXmlMixin):
         self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-submodule-advanced'), factory.app.create_suite())
 
     def test_case_list_form_parent_child_submodule_advanced_rename_case_var(self):
-        """Test that the session vars in the entries for the submodule get updated
-        to match the parent (and to avoid naming conflicts).
-        m3-f0: 'case_id_load_house' -> 'case_id_load_house_renamed'
-        m3-f0: 'case_id_load_house_renamed' -> 'case_id_load_house_renamed_person'
-        """
+        # Test that the session vars in the entries for the submodule get updated
+        # to match the parent (and to avoid naming conflicts).
+        # m3-f0: 'case_id_load_house' -> 'case_id_load_house_renamed'
+        # m3-f0: 'case_id_load_house_renamed' -> 'case_id_load_house_renamed_person'
         factory = AppFactory(build_version='2.9')
         register_house_module, register_house_form = factory.new_basic_module('register_house', 'house')
         factory.form_opens_case(register_house_form)
@@ -276,16 +267,14 @@ class CaseListFormSuiteTests(SimpleTestCase, TestXmlMixin):
         self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-submodule-advanced-rename-var'), factory.app.create_suite())
 
     def test_case_list_form_parent_child_submodule_mixed(self):
-        """
-        * Register house (case type = house, basic)
-          * Register house form
-        * Register person (case type = person, parent select = 'Register house', advanced)
-          * Register person form
-        * Update house (case type = house, case list form = 'Register house')
-          * Update house form
-          * Update person (case type = person, case list form = 'Register person form', advanced, parent module = 'Update house')
-              * Update person form
-        """
+        # * Register house (case type = house, basic)
+        #   * Register house form
+        # * Register person (case type = person, parent select = 'Register house', advanced)
+        #   * Register person form
+        # * Update house (case type = house, case list form = 'Register house')
+        #   * Update house form
+        #   * Update person (case type = person, case list form = 'Register person form', advanced, parent module = 'Update house')
+        #       * Update person form
         factory = AppFactory(build_version='2.9')
         register_house_module, register_house_form = factory.new_basic_module('register_house', 'house')
         factory.form_opens_case(register_house_form)
@@ -314,19 +303,16 @@ class CaseListFormSuiteTests(SimpleTestCase, TestXmlMixin):
         self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-submodule-mixed'), factory.app.create_suite())
 
     def test_target_module_different_datums(self):
-        """
-        * Registration
-          * Register patient form
-            * open case (patient), update user_case
-        * Visits (case type = patient, case list form = 'Register patient')
-          * Visit form
-            * update case, open child case (visit), load from user_case
-          * Record notes
-            * update case, open child case (visit)
-          * Update patient
-            * update case
-        """
-
+        # * Registration
+        #   * Register patient form
+        #     * open case (patient), update user_case
+        # * Visits (case type = patient, case list form = 'Register patient')
+        #   * Visit form
+        #     * update case, open child case (visit), load from user_case
+        #   * Record notes
+        #     * update case, open child case (visit)
+        #   * Update patient
+        #     * update case
         factory = AppFactory(build_version='2.9')
         registration_module, registration_form = factory.new_basic_module('registration', 'patient')
 

--- a/corehq/apps/app_manager/tests/util.py
+++ b/corehq/apps/app_manager/tests/util.py
@@ -113,8 +113,8 @@ def _check_shared(expected, actual, checker, extension):
     if not checker.check_output(expected, actual, 0):
         original_message = message = "{} mismatch\n\n".format(extension.upper())
         diff = difflib.unified_diff(
-            actual.splitlines(1),
             expected.splitlines(1),
+            actual.splitlines(1),
             fromfile='want.{}'.format(extension),
             tofile='got.{}'.format(extension)
         )

